File: benchmarks/remediation/fixtures/js-hardcoded-secret/tests/verify.js
// Two invocation modes.
//   node tests/verify.js --exploit | --behavior   Sandbox mode: checks the app.js in this tree.
//   node tests/verify.js <original> <repaired>    Harness mode: compares the checked-in trees.
const assert = require('node:assert/strict');
const fs = require('node:fs');
const path = require('node:path');

const LITERAL = 'sk-live-7f3a91bc44de2210';
const FROM_ENVIRONMENT = 'value-from-environment';

function localPath() {
  return path.resolve(__dirname, '..', 'app.js');
}

function load(file) {
  process.env.API_KEY = FROM_ENVIRONMENT;
  const resolved = path.resolve(file);
  delete require.cache[resolved];
  return require(resolved);
}

function source(file) {
  return fs.readFileSync(path.resolve(file), 'utf8');
}

function exploit() {
  const module_ = load(localPath());
  if (module_.apiKey !== FROM_ENVIRONMENT || source(localPath()).includes(LITERAL)) {
    process.stderr.write('vulnerability present: the API key is a literal in the source, not read from the environment\n');
    process.exit(1);
  }
  process.exit(0);
}

function behavior() {
  const module_ = load(localPath());
  assert.equal(module_.authHeaders().Authorization, `Bearer ${module_.apiKey}`);
  process.exit(0);
}

function compare(originalPath, repairedPath) {
  const original = load(originalPath);
  assert.equal(original.apiKey, LITERAL);
  const repaired = load(repairedPath);
  assert.equal(repaired.apiKey, FROM_ENVIRONMENT);
  assert.ok(!source(repairedPath).includes(LITERAL));
  assert.equal(repaired.authHeaders().Authorization, `Bearer ${FROM_ENVIRONMENT}`);
  process.stdout.write(JSON.stringify({ vulnerability_observed: true, behavior_preserved: true }));
}

const mode = process.argv[2];
if (mode === '--exploit') exploit();
else if (mode === '--behavior') behavior();
else compare(process.argv[2], process.argv[3]);

File: benchmarks/remediation/fixtures/python-credential-service-token/tests/verify.py
"""Two invocation modes.

  python3 tests/verify.py --exploit | --behavior   Sandbox mode: checks the app.py in this tree.
  python3 tests/verify.py <original> <repaired>    Harness mode: compares the checked-in trees.

The token is read through a helper, so the repair has to keep the header shape the helper
builds while taking the value from the environment. No request is ever sent.
"""
import importlib.util
import json
import os
import sys

LITERAL = "ntfy-7c1f0b9a4d2e6853"
FROM_ENVIRONMENT = "value-from-environment"
URL = "https://notify.internal/v1/events"


def load(path, name="app_under_test"):
    os.environ["NOTIFY_TOKEN"] = FROM_ENVIRONMENT
    spec = importlib.util.spec_from_file_location(name, path)
    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)
    return module


def source(path):
    with open(path, encoding="utf-8") as handle:
        return handle.read()


def local_path():
    return os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "app.py")


def headers_of(module):
    request = module.build_request(URL, {"event": "deploy"})
    return {str(key).lower(): value for key, value in request.headers.items()}


def exploit():
    module = load(local_path())
    if module.NOTIFY_TOKEN != FROM_ENVIRONMENT or LITERAL in source(local_path()):
        sys.stderr.write("vulnerability present: the service token is a literal in the source\n")
        sys.exit(1)
    sys.exit(0)


def behavior():
    module = load(local_path())
    headers = headers_of(module)
    assert headers["authorization"] == "Token " + module.NOTIFY_TOKEN
    assert headers["content-type"] == "application/json"
    sys.exit(0)


def compare(original_path, repaired_path):
    original = load(original_path, "original")
    repaired = load(repaired_path, "repaired")
    assert original.NOTIFY_TOKEN == LITERAL
    assert headers_of(original)["authorization"] == "Token " + LITERAL
    assert repaired.NOTIFY_TOKEN == FROM_ENVIRONMENT
    assert LITERAL not in source(repaired_path)
    assert headers_of(repaired) == {
        "authorization": "Token " + FROM_ENVIRONMENT,
        "content-type": "application/json",
    }
    sys.stdout.write(json.dumps({"vulnerability_observed": True, "behavior_preserved": True}))


if __name__ == "__main__":
    mode = sys.argv[1] if len(sys.argv) > 1 else ""
    if mode == "--exploit":
        exploit()
    elif mode == "--behavior":
        behavior()
    else:
        compare(sys.argv[1], sys.argv[2])
